rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -470,6 +483,11 @@ NodeParameters::set_parameters_atomically(const std::vector<rclcpp::Parameter> &
 {
   std::lock_guard<std::recursive_mutex> lock(mutex_);
 
+  if (!parameter_modification_enabled_) {
+    throw rclcpp::exceptions::ParameterModifiedInCallbackException(
+            "cannot modify a parameter from within set callback");
+  }
+
```

Comment:
I think we can add here:
```
parameter_modification_enabled_ = false;
RCLCPP_SCOPE_EXIT({parameter_modification_enabled_ = true;});
```
and avoid similar code in each call to `__*_common` within the method. Is that correct?

We can maybe apply that style in the other methods, keeping that code after the 
```cpp
if (!parameter_modification_enabled_) {
``` 
in `declare_parameter` method too.
I think it's a bit clearer.

Reply:
Oh, I like it.  This keeps all of the code that deals with `parameter_modification_enabled_` in the same place.  It has the slight downside in that it is holding the "lock" for a longer period of time, but that doesn't seem to matter in my testing so far.  I'm going to do a bit more testing here, and assuming that is the case, there is no need to touch `__set_parameters_atomically` or `__declare_parameters`.  Thanks for the idea!